Makefile

```diff
@@ -1,22 +1,127 @@
-.PHONY: help clean spec mystspec
+# Because env `SHELL` might be 
+# some extremely weird shell
+SHELL = bash
 
-spec:      ## Runs all specs
-	crystal spec
-	crystal run src/myst_cli.cr -- spec/myst/spec.mt
+# Parse makefiles named *.mk in the dir `local`
+# for local extra makefile targets if wanted.
+# It is smart to set MYSTBUILD to your myst dev-build there
+-include local/*.mk
 
-myst-spec: ## Runs just the in-language specs
-	crystal run src/myst_cli.cr -- spec/myst/spec.mt
-	
-build:     ## Builds myst into an executable
-	shards build
+# https://www.gnu.org/software/make/manual/html_node/One-Shell.html
+.ONESHELL:
 
-check:     ## Runs all crystal specs
-	crystal spec/
+# The docs says i should do this
+.SUFFIXES: .cr .mt
 
-clean:     ## Cleans (deletes) docs and executables
-	rm -rf docs
-	rm bin/*
+MYST_CLI = src/myst_cli.cr
 
-# https://gist.github.com/prwhite/8168133 "Help target"
-help:      ## Show this help.
-	@fgrep -h "##" $(MAKEFILE_LIST) | fgrep -v fgrep | sed -e 's/\\$$//' | sed -e 's/##//'
+MYST_INTERPRETER_SPEC = spec/all_spec.cr
+MYST_IN_LANG_SPEC 		= spec/myst/spec.mt
+
+SOURCE_FILES := $(shell find src    -type f -name '*.cr')
+SPEC_FILES   := $(shell find spec   -type f -name '*.cr' -o -name '*.mt')
+STDLIB_FILES := $(shell find stdlib -type f -name '*.mt')
+ALL_FILES    := $(SPEC_FILES) $(STDLIB_FILES) $(SOURCE_FILES)
+
+INSTALL_LOCATION ?= /usr/local/bin/myst
+MYSTBUILD				 ?= myst
+
+O ?= bin
+
+info ?= true
+
+# If another default is intentionally wished for,
+# set it in local/*.mk, like this:
+# .DEFAULT_GOAL = goal
+# 
+.DEFAULT_GOAL ?= spec
+
+info_log = $(if $(subst false,,$(info)),$(info $(1)))
+
+# Makefile convention 
+all: $(O)/myst $(O)/spec
+
+.PHONY: spec
+spec: $(O)/myst $(O)/spec ## Runs all specs
+	$(call info_log,Running interpreter spec)
+	$(O)/spec
+	$(call info_log,Running in-language spec)
+	$(O)/myst $(MYST_IN_LANG_SPEC)
+
+myst-spec: $(O)/myst ## Runs just the in-language specs
+	$(O)/myst $(MYST_IN_LANG_SPEC)
+
+.PHONY: install
+install: $(INSTALL_LOCATION) ## Install myst to INSTALL_LOCATION		
+
+.PHONY: build
+build: $(O)/myst ## Builds myst into an executable
+
+.PHONY: myst-spec_with_build
+myst-spec_with_build: ## Runs the in-language specs with MYSTBUILD
+	$(MYSTBUILD) $(MYST_IN_LANG_SPEC)
+
+$(O)/myst: $(SOURCE_FILES)
+	$(call info_log,Building myst...)
+	mkdir -p $(O)
+	crystal build -o $(O)/myst $(MYST_CLI)
```

Comment:
I think using `shards build` for this might be more consistent, rather than relying on `$(O)` to point to the right place. Not a requirement, just a thought :)

Reply:
I thought about that. `OUT` (which is what i'll change `O` to) can be set like this `make build OUT=.../somewhere`, so im pretty sure `OUT` will be the right place, and `$(OUT)/myst` is required to be a file for the `$(SOURCE_FILES)` dependency to work.

Reply:
So if you want `shards build` then just use `shards build` not `make build` is my point. :)